### PR TITLE
Apigroup and alpha resource for petset

### PR DESCRIPTION
**Pull request by bprashanth**
I'm still writing unittests, validation and comments but I'm planning to punt on anything more involved since this is alpha. 

@smarterclayton @bgrant0607 @thockin probably only care about first commit. I'm hoping we can take this as a rough cut and iterate. 

@lavalamp for general api machinery

### CODE REVIEW

**Comment by bprashanth** [comment]
This is ready for review. It'll fail verify-api-refdocs because of: https://github.com/kubernetes/kubernetes/issues/24310. I've tried to keep the first cut as similar to the RC as I can.

File: pkg/apis/apps/v1alpha1/types.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+	Name string `json:"name,omitempty"`
+
+	// Volumes are the volumes associated with this pet.
+	Volumes []v1.Volume `json:"volumes,omitempty"`
+
+	// Role is the role of this pet in the cluster.
+	// All pets start off as non-members of an existing PetSet cluster. They
+	// become members when they flip to "Ready". This field might be updated
+	// by a babysitter controller that has more clarity about roles.
+	Role string `json:"role,omitempty"`
```
Comment (smarterclayton) [inline pkg/apis/apps/v1alpha1/types.go]: Should be a conditions array.

File: pkg/apis/apps/v1alpha1/types.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+	ServiceName string `json:"serviceName"`
+}
+
+// PetStatus represents the current state of a single pet in the set.
+type PetStatus struct {
+	// Name is the name of this particular pet.
+	Name string `json:"name,omitempty"`
+
+	// Volumes are the volumes associated with this pet.
+	Volumes []v1.Volume `json:"volumes,omitempty"`
```
Comment (smarterclayton) [inline pkg/apis/apps/v1alpha1/types.go]: Why is this necessary?

File: pkg/apis/apps/install/install.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
+/*
```
Comment (smarterclayton) [inline pkg/apis/apps/install/install.go]: Not a fan of `apps`, since lots of apps won't be pet sets.  `clustered` is too close to the `cluster` resource.

File: pkg/apis/apps/v1alpha1/types.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+	// Role is the role of this pet in the cluster.
+	// All pets start off as non-members of an existing PetSet cluster. They
+	// become members when they flip to "Ready". This field might be updated
+	// by a babysitter controller that has more clarity about roles.
+	Role string `json:"role,omitempty"`
+}
+
+// PetSetStatus represents the current state of a PetSet.
+type PetSetStatus struct {
+	Pets []PetStatus `json:"pets,omitempty"`
```
Comment (smarterclayton) [inline pkg/apis/apps/v1alpha1/types.go]: Do we actually need this?  In the proposal we stated we don't need to do this, I'd prefer the sets to be deterministic based on name and label selection and brian weighed in as not wanting to identify the set.

File: pkg/apis/apps/v1alpha1/types.go (hunk 2 above)
Comment (smarterclayton) [inline pkg/apis/apps/v1alpha1/types.go]: Should just be a naming pattern.

File: pkg/apis/apps/v1alpha1/types.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+	Template v1.PodTemplateSpec `json:"template"`
+
+	// VolumeClaims is a list of claims that pets are allowed to reference.
+	// The PetSet controller is responsible for mapping network identities to
+	// claims in a way that maintains the identity of a pet. Every claim in
+	// this list must have at least one matching (by name) volumeMount in one
+	// container in the template. A claim in this list takes precedence over
+	// any volumes in the template, with the same name.
+	// TODO: Define the behavior if a claim already exists with the same name.
+	VolumeClaims []v1.PersistentVolumeClaim `json:"volumeClaims,omitempty"`
```
Comment (smarterclayton) [inline pkg/apis/apps/v1alpha1/types.go]: `VolumeClaimTemplates`

File: pkg/apis/apps/v1alpha1/types.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+	// ServiceName is the name of the service that governs this PetSet.
```
Comment (smarterclayton) [inline pkg/apis/apps/v1alpha1/types.go]: Might want to say that this is the service that provides network identity for this pet set.

File: pkg/apis/apps/v1alpha1/types.go [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+	Status PetSetStatus `json:"status,omitempty"`
+}
+
+// A PetSetSpec is the specification of a PetSet.
+type PetSetSpec struct {
+	// Replicas is the desired number of replicas of the gieven Template.
+	// These are replicas in the sense that they are instantiations of the
+	// same Template, but individual replicas also have a consistent identity.
+	// TODO: Consider a rename of this field.
+	Replicas int `json:"replicas"`
```
Comment (smarterclayton) [inline pkg/apis/apps/v1alpha1/types.go]: Pointer, defaults to 1.

File: pkg/apis/apps/v1alpha1/types.go [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+	// Selector is a label query over pods that should match the replica count.
+	// If empty, defaulted to labes on the pod template.
```
Comment (smarterclayton) [inline pkg/apis/apps/v1alpha1/types.go]: Typo

File: pkg/apis/apps/v1alpha1/types.go (hunk 4 above)
Comment (bprashanth) [inline pkg/apis/apps/v1alpha1/types.go]: We can remove it, I imagined describe as showing me which DNS names mapped to which volumes of what type and capacity, but I'm fine leaving it out for now. Just make this replicas?

File: pkg/apis/apps/v1alpha1/types.go (hunk 2 above)
Comment (bprashanth) [inline pkg/apis/apps/v1alpha1/types.go]: I figured giving more information about the volumes allocated (name+type+capacity) in describe would be better, since the petsest only includes a list of claims. I can make it a list of strings if you prefer?

File: pkg/apis/apps/install/install.go (hunk 3 above)
Comment (bprashanth) [inline pkg/apis/apps/install/install.go]: I think the idea is to get daemon,replica etc out of "extensions". Open to suggestion, apps is just something we came up with after ~10m of discussions, and no one really had anything better.

File: pkg/apis/apps/v1alpha1/types.go (hunk 2 above)
Comment (smarterclayton) [inline pkg/apis/apps/v1alpha1/types.go]: Would that not just be a label selector on the claims created?  I'd
imagined:

kubectl get petset
  petset-mine

kubectl get pods
  petset-mine-1
  petset-mine-2
  petset-mine-3

kubectl get pvc
  petset-mine-1-1
  petset-mine-1-2
  petset-mine-2-1
  petset-mine-2-2
  petset-mine-3-1
  petset-mine-3-2

with labels being applied to the pods and pvcs for grouping.  If you had
pet set size 100 (a reasonable high end for now) I'd rather have the volume
names be deterministic rather than try to show the list.  Controller only
has to fill in gaps, babysitter could use annotations on the petset for now
for locking if necessary (or on the service endpoints via discovery on the
serviceRef)

File: pkg/apis/apps/install/install.go (hunk 3 above)
Comment (smarterclayton) [inline pkg/apis/apps/install/install.go]: Yeah agree with desire to create its own group, I definitely want a
different name before we merge (will think hard on this).

File: pkg/apis/apps/v1alpha1/types.go (hunk 2 above)
Comment (bprashanth) [inline pkg/apis/apps/v1alpha1/types.go]: SG

**Comment by bprashanth** [comment]
Trap is green. Very optimistic request for LGTM, it's going in alpha so we can iterate?

I'm going with "apps" over "app" since we have "extensions" and "metrics". I like Clustered but it's too close to many things. Haven't heard a better suggestion, but I can always send a follow-up.

**Comment by smarterclayton** [comment]
"sets.PetSet"?  "consensus.PetSet"?  "systems.PetSet"?
 "managedsets.PetSet"?  I could see something with managed - but for now
"apps" is ok.  I just don't want someone to have "apps.Simple" or
"apps.App" or "apps.Microservice" and overlap in here.

File: pkg/apis/apps/v1alpha1/types_swagger_doc_generated.go [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+//
+// TODOs are ignored from the parser (e.g. TODO(andronat):... || TODO:...) if and only if
+// they are on one line! For multiple line or blocks that you want to ignore use ---.
+// Any context after a --- is ignored.
+//
+// Those methods can be generated by using hack/update-generated-swagger-docs.sh
+
+// AUTO-GENERATED FUNCTIONS START HERE
+var map_PetSet = map[string]string{
+	"":       "PetSet represents a set of pods with consistent identities. Identities are defined as:\n - Network: A single stable DNS and hostname.\n - Storage: As many VolumeClaims as requested.\nThe PetSet guarantees that a given network identity will always map to the same storage identity.",
```
Comment (smarterclayton) [inline pkg/apis/apps/v1alpha1/types_swagger_doc_generated.go]: Should we doc PetSet as being experimental and subject to change?

File: pkg/registry/petset/etcd/etcd_test.go [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+import (
+	"testing"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/apis/apps"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/registry/generic"
+	"k8s.io/kubernetes/pkg/registry/registrytest"
```
Comment (smarterclayton) [inline pkg/registry/petset/etcd/etcd_test.go]: Add the resttests

Comment (smarterclayton) [inline pkg/registry/petset/etcd/etcd_test.go]: Or did we rename resttest to registrytest and I missed it?

File: pkg/registry/petset/strategy.go [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+// watch events from etcd to clients of the apiserver only interested in specific
+// labels/fields.
+func MatchPetSet(label labels.Selector, field fields.Selector) generic.Matcher {
+	return &generic.SelectionPredicate{
+		Label: label,
+		Field: field,
+		GetAttrs: func(obj runtime.Object) (labels.Set, fields.Set, error) {
+			petSet, ok := obj.(*apps.PetSet)
+			if !ok {
+				return nil, nil, fmt.Errorf("Given object is not an PetSet.")
```
Comment (smarterclayton) [inline pkg/registry/petset/strategy.go]: Lowercase leading word

**Comment by smarterclayton** [comment]
A few comments

File: pkg/registry/petset/etcd/etcd_test.go (hunk 10 above)
Comment (bprashanth) [inline pkg/registry/petset/etcd/etcd_test.go]: Yeah the tester used in registrytest is resttest now (https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/registrytest/etcd.go#L53), assuming that's what you ment.

File: pkg/apis/apps/v1alpha1/types_swagger_doc_generated.go (hunk 9 above)
Comment (bprashanth) [inline pkg/apis/apps/v1alpha1/types_swagger_doc_generated.go]: Might as well

**Comment by bprashanth** [comment]
@k8s-bot test this github issue: #24451

**Comment by bprashanth** [comment]
PTAL

**Comment by smarterclayton** [comment]
So fishing around on names, "apps" has raised some eyebrows.

Would we put "ReplicaSets" into "apps"?  That's how I build my web apps.  Would we put "Services" into "apps"?  I can't build an app without a service.  How about DaemonSets?  Are they things I can use to run "cluster level apps"?

I still don't have a better suggestion, but I'd almost prefer to pick a bad name here (like "petset") than something that is halfway plausible.

**Comment by bprashanth** [comment]
I view this as: everything you need to run your app outside kube is in apps/. This includes replicas, state/identity, updates. I don't view a Service as part of your app, as in you give me django, i put it in a pod, run N replicas of your app (stateful or stateless) and expose it to the rest of the system/externally through a Service. If I wasn't on kube I would just use ports + ingress, I'd still run replicas, deploy..
DaemonSets are also not apps IMO.

I feel like just "sets" was probably vetoed before?

**Comment by bprashanth** [comment]
Also networking and storage are not a *part* of your app in that they're swappable, the contents of an RC/petset/deployment can't be backed by anything else (same for daemon, but I'm assuming no one writing apps is writing daemons -- though that doesn't appear to be the case from user channels).

**Comment by smarterclayton** [comment]
Ok, I'm fine with apps for now, we can revisit when it comes out of alpha.  Squash and I'll LGTM

**Comment by bprashanth** [comment]
Squashed and green, PTAL

File: pkg/apis/apps/types.go [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+	Spec PetSetSpec `json:"spec,omitempty"`
+
+	// Status is the current status of Pets in this PetSet. This data
+	// may be out of date by some window of time.
+	Status PetSetStatus `json:"status,omitempty"`
+}
+
+// A PetSetSpec is the specification of a PetSet.
+type PetSetSpec struct {
+	// Replicas is the desired number of replicas of the gieven Template.
```
Comment (thockin) [inline pkg/apis/apps/types.go]: given

File: pkg/apis/apps/validation/validation.go [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+func ValidatePetSetSpec(spec *apps.PetSetSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	allErrs = append(allErrs, apivalidation.ValidateNonnegativeField(int64(spec.Replicas), fldPath.Child("replicas"))...)
+	if spec.Selector == nil {
+		allErrs = append(allErrs, field.Required(fldPath.Child("selector"), ""))
+	} else {
+		allErrs = append(allErrs, unversionedvalidation.ValidateLabelSelector(spec.Selector, fldPath.Child("selector"))...)
+		if len(spec.Selector.MatchLabels)+len(spec.Selector.MatchExpressions) == 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("selector"), spec.Selector, "empty selector is not valid for deployment."))
```
Comment (thockin) [inline pkg/apis/apps/validation/validation.go]: s/deployment/petset

File: pkg/apis/apps/validation/validation.go [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+		}
+	}
+
+	selector, err := unversioned.LabelSelectorAsSelector(spec.Selector)
+	if err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("selector"), spec.Selector, "invalid label selector."))
```
Comment (thockin) [inline pkg/apis/apps/validation/validation.go]: you don't need to say "invalid label selector" - the error will already say "invalid" and "selector".  The extra field is just for extra details, which you don't have

File: pkg/registry/cachesize/cachesize.go [hunk 15]
```diff
@@ -34,6 +34,7 @@ const (
 	Endpoints                Resource = "endpoints"
 	HorizontalPodAutoscalers Resource = "horizontalpodautoscalers"
 	Ingress                  Resource = "ingress"
+	PetSet                   Resource = "petset"
```
Comment (thockin) [inline pkg/registry/cachesize/cachesize.go]: We *really* need to stop open-coding this stuff all over.

File: pkg/registry/petset/strategy.go [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+func (petSetStrategy) Canonicalize(obj runtime.Object) {
+}
+
+// AllowCreateOnUpdate is false for PetSet; this means POST is needed to create one.
+func (petSetStrategy) AllowCreateOnUpdate() bool {
+	return false
+}
+
+// ValidateUpdate is the default update validation for an end user.
+func (petSetStrategy) ValidateUpdate(ctx api.Context, obj, old runtime.Object) field.ErrorList {
```
Comment (thockin) [inline pkg/registry/petset/strategy.go]: Are we allowing update?  Lesson from past objects - disable it to start and selectively enable some fields to be updated...

Comment (bprashanth) [inline pkg/registry/petset/strategy.go]: How would we disallow an update to replicas, selectors, labels, annotations etc? there's a petsets/status subresource.

Comment (thockin) [inline pkg/registry/petset/strategy.go]: Are you asking how technically or how logistically?  Technically, just
return an error

Comment (bprashanth) [inline pkg/registry/petset/strategy.go]: I mean that would make a lame resource :)

Comment (smarterclayton) [inline pkg/registry/petset/strategy.go]: +1 we should put that on conventions or API conventions or API changes

Comment (bprashanth) [inline pkg/registry/petset/strategy.go]: confused, are you saying don't allow updates?

Comment (thockin) [inline pkg/registry/petset/strategy.go]: Start with the assumption that all fields are immutable and selectively enable updates for fields ONLY after careful consideration

Comment (bprashanth) [inline pkg/registry/petset/strategy.go]: SG

**Comment by bprashanth** [comment]
PTAL, tests passed. 
I've disallowed udpates to spec, fixed nits and added a unittests for update through the status subresource.

**Comment by thockin** [comment]
I gave it a skim, and it looks good.  I will try to do a deeper review if nobody gets there before me.

**Comment by smarterclayton** [comment]
I have no further comments.

**Comment by bprashanth** [comment]
@thockin can i inherit previous lgtm?

**Comment by thockin** [comment]
No release note?  when do we add a release note?

**Comment by smarterclayton** [comment]
Probably when there's actually something implemented.  I would expect
slightly later in the release (since there's no controller yet)

**Comment by bprashanth** [comment]
When we're confident it adds value that we can define in words included in the note.

**Comment by smarterclayton** [comment]
Whoa - `k8s-merge-robot merged commit 1f8bd21 into kubernetes:master 2 of 6 checks passed` is not good.

**Comment by thockin** [comment]
@eparis investigate the 2 of 6?

**Comment by eparis** [comment]
https://github.com/kubernetes/contrib/issues/422

rebases can race with the re-test....
